server/join_test.go

```diff
@@ -85,33 +78,274 @@ func waitMembers(svr *Server, c int) error {
 	return errors.New("waitMembers Timeout")
 }
 
-func (s *testJoinServerSuite) TestJoin(c *C) {
-	svrs := make([]*Server, 0, len(s.cfgs))
-	for i, cfg := range s.cfgs {
-		svr, err := NewServer(cfg)
+func waitLeader(svrs []*Server) error {
+	// maxRetryTime * waitInterval = 10s
+	maxRetryCount := 20
+	waitInterval := 500 * time.Millisecond
+	for count := 0; count < maxRetryCount; count++ {
+		for _, s := range svrs {
+			// TODO: a better way of finding leader.
+			if s.etcd.Server.Leader() == s.etcd.Server.ID() {
+				return nil
+			}
+		}
+		time.Sleep(waitInterval)
+	}
+	return errTimeout
+}
+
+// notice: cfg has changed
+func startPdWith(cfg *Config) (*Server, error) {
+	svrCh := make(chan *Server)
+	errCh := make(chan error)
+	go func() {
+		svr, err := CreateServer(cfg)
+		if err != nil {
+			errCh <- errors.Trace(err)
+			return
+		}
+		err = svr.StartEtcd(nil)
+		if err != nil {
+			errCh <- errors.Trace(err)
+			return
+		}
+
+		svrCh <- svr
+		svr.Run()
+	}()
+
+	timer := time.NewTimer(5 * time.Second)
+	defer timer.Stop()
+
+	select {
+	case s := <-svrCh:
+		return s, nil
+	case e := <-errCh:
+		return nil, errors.Trace(e)
+	case <-timer.C:
+		return nil, errTimeout
+	}
+}
+
+type cleanUpFunc func()
+
+func mustNewJoinCluster(c *C, num int) ([]*Config, []*Server, cleanUpFunc) {
+	svrs := make([]*Server, 0, num)
+	cfgs := newTestMultiJoinConfig(num)
+
+	for _, cfg := range cfgs {
+		svr, err := startPdWith(cfg)
 		c.Assert(err, IsNil)
-		defer svr.Close()
 		svrs = append(svrs, svr)
+	}
 
-		go svr.Run()
+	waitMembers(svrs[rand.Intn(num)], num)
 
-		// Make sure new pd is started.
-		err = waitMembers(svrs[0], i+1)
-		c.Assert(err, IsNil)
+	// clean up
+	clean := func() {
+		for _, s := range svrs {
+			s.Close()
+		}
+		for _, c := range cfgs {
+			os.RemoveAll(c.DataDir)
+		}
 	}
 
-	endpoints := strings.Split(s.cfgs[rand.Intn(3)].ClientUrls, ",")
-	client, err := clientv3.New(clientv3.Config{
-		Endpoints:   endpoints,
-		DialTimeout: 3 * time.Second,
-	})
+	return cfgs, svrs, clean
+}
+
+func alive(target, peer *Server) error {
```

Comment:
"alive" seems misleading, it is actually testing whether these two servers are in the same cluster and they are connective?